### PR TITLE
feat: reorgs outside of confirmations range handling

### PR DESCRIPTION
**Pull request by nduchak**

### CODE REVIEW

File: src/cli/start.ts [hunk 1]
```diff
@@ -103,23 +97,42 @@ ${formattedServices}`
...
+      // reorgs outside of confirmation range.
+      await resetPromise
+
+      logger.warn('Reorg detected outside of confirmation range. Rebuilding the service\'s state!')
+      logger.info('Stopping service')
+      stopCallback()
+      backUpJob.stop()
+
+      // Restore DB from backup
+      await backUpJob.restoreDb(() => undefined)
```
Comment (AuHau) [inline src/cli/start.ts:130]: Why do you "ignore" the `errorCallback` here? And actually why there is even such a thing? Isn't rejected Promise enough?

**Review by AuHau — CHANGES_REQUESTED** [review]
Hmm there are few questions and some things to improve. But it is on good path!

File: src/cli/start.ts [hunk 2]
```diff
@@ -103,23 +97,42 @@ ${formattedServices}`
     const configOverrides = this.buildConfigObject(flags)
     config.util.extendDeep(config, configOverrides)
 
-    if (flags.purge) {
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/cli/start.ts:106]: Will do

File: src/cli/start.ts [hunk 3]
```diff
@@ -103,23 +97,42 @@ ${formattedServices}`
...
+    const backUpJob = new DbBackUpJob(ethFactory())
+    // An infinite loop which you can exit only with SIGINT/SIGKILL
+    while (true) {
+      let stopCallback: () => void = () => {
+        throw new Error('No stop callback was assigned!')
+      }
 
-    const app = await appFactory()
-    const port = config.get('port')
-    const server = app.listen(port)
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/cli/start.ts:113]: Just want to make the same as we had it in pinner, as we use the same algorithm for reorgs

File: src/services/storage/index.ts [hunk 4]
```diff
@@ -110,6 +113,13 @@ const storage: CachedService = {
     })
     eventsEmitter.on('newConfirmation', (data) => confirmationService.emit('newConfirmation', data))
     eventsEmitter.on('invalidConfirmation', (data) => confirmationService.emit('invalidConfirmation', data))
+    eventsEmitter.on(REORG_OUT_OF_RANGE_EVENT_NAME, (blockNumber: number) => reorgEmitterService.emitReorg(blockNumber, 'storage'))
+
+    return {
+      stop: () => {
+        eventsEmitter.stop()
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/services/storage/index.ts:120]: True 👍

File: src/utils.ts [hunk 5]
```diff
@@ -196,3 +208,145 @@ export abstract class BaseCLICommand extends Command {
...
+      .sort(
+        (a: Record<string, any>, b: Record<string, any>) =>
+          a.block.number.gt(b.block.number) ? -1 : 1
+      )
+  }
+
+  return []
+}
+
+export class DbBackUpJob {
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/utils.ts:241]: Good idea

File: src/utils.ts [hunk 6]
```diff
@@ -196,3 +208,145 @@ export abstract class BaseCLICommand extends Command {
...
+  return {
+    name: backUpName,
+    block: { number: new BigNumber(blockNumber), hash }
+  }
+}
+
+function getBackUps (): BackUpEntry[] {
+  const backupConfig = config.get<DbBackUpConfig>('dbBackUp')
+
+  const backups = fs.readdirSync(path.resolve(__dirname, `../${backupConfig.path}`))
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/utils.ts:227]: Will do

File: src/blockchain/utils.ts [hunk 7]
```diff
@@ -29,7 +28,7 @@ export function isServiceInitialized (serviceName: string): boolean {
   return blockTracker.getLastFetchedBlock()[0] !== undefined
 }
 
-export function getEventsEmitterForService (serviceName: string, eth: Eth, contractAbi: AbiItem[]): EventEmitter {
+export function getEventsEmitterForService (serviceName: string, eth: Eth, contractAbi: AbiItem[]): BaseEventsEmitter {
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/blockchain/utils.ts:31]: I need the `stop` method from `BaseEvemtsEmitter`

File: src/definitions.ts [hunk 8]
```diff
@@ -28,7 +29,8 @@ export enum ServiceAddresses {
   STORAGE_AGREEMENTS = '/storage/v0/agreements',
   XR = '/rates/v0/',
   CONFIRMATIONS = '/confirmations',
-  NEW_BLOCK_EMITTER = '/new-block'
+  NEW_BLOCK_EMITTER = '/new-block',
+  REORG_EMITTER = '/reorg'
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/definitions.ts:33]: I think it's fine as we can have different events based on this path. Separation happens there

File: src/utils.ts [hunk 9]
```diff
@@ -196,3 +208,145 @@ export abstract class BaseCLICommand extends Command {
     return Promise.resolve()
   }
 }
+
+type BackUpEntry = { name: string, block: { hash: string, number: BigNumber } }
+
+function parseBackUps (backUpName: string): BackUpEntry {
+  const [block] = backUpName.split('.')[0].split('-')
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/utils.ts:215]: Not too big, like:  add-name-to-user

File: test/cli.spec.ts [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    const dbPath = path.join(__dirname, '..', '..', config.get<string>('db'))
+    try {
+      unlinkSync(dbPath)
+    } catch (e) {
+      // Ignore "not found" errors
+      if (e.code !== 'ENOENT') {
+        throw e
+      }
+    }
+    process.env.RIFS_DB = dbPath
```
Comment (AuHau) [inline test/cli.spec.ts:41]: I see you copy&pasted a test from the Pinner. This won't work as the env. variable is `RIFM_DB` here. But tbh. not sure if it is needed? 🤔

File: test/db-backup-restore.spec.ts [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+      const eth = Substitute.for<Eth>()
+      const backupJob = new DbBackUpJob(eth)
+
+      expect(fs.readdirSync(config.get<DbBackUpConfig>('dbBackUp').path).length).to.be.eql(0)
+      await expect(backupJob.restoreDb()).to.eventually.be.rejectedWith(
+        Error,
+        'Should be two backups to be able to restore'
+      )
+    })
+    it('should throw if backup block hash is not exist after reorg', async () => {
```
Comment (AuHau) [inline test/db-backup-restore.spec.ts:35]: Can you please add test-case for when you get a different hash (eq. the reorg is bigger then the backup)?

File: src/cli/start.ts [hunk 12]
```diff
@@ -103,23 +94,42 @@ ${formattedServices}`
...
+      // reorgs outside of confirmation range.
+      await resetPromise
+
+      logger.warn('Reorg detected outside of confirmation range. Rebuilding the service\'s state!')
+      logger.info('Stopping service')
+      stopCallback()
+      backUpJob.stop()
+
+      // Restore DB from backup
+      await backUpJob.restoreDb()
```
Comment (AuHau) [inline src/cli/start.ts:127]: Hmmm so if the restore procedure fails for some reason then Cache craches, right? 🤔

File: src/cli/start.ts [hunk 13]
```diff
@@ -103,23 +94,42 @@ ${formattedServices}`
...
 
-    server.on('listening', () =>
-      this.log('Server started on port %d', port)
-    )
+      // Let see if we have to restart the app at some point most probably because
+      // reorgs outside of confirmation range.
+      await resetPromise
+
+      logger.warn('Reorg detected outside of confirmation range. Rebuilding the service\'s state!')
+      logger.info('Stopping service')
```
Comment (AuHau) [inline src/cli/start.ts:122]:
```suggestion
      logger.info('Stopping the app')
```

**Review by AuHau — CHANGES_REQUESTED** [review]
Hmm still some things to tackle. 

Btw. did you test how the UI handles when Cache disconnects? We should know its behavior as maybe there should be some added logic to the UI for handling when Cache is not online and also to listen on the Reorg channel...

Also, please don' do force-push once the review process starts. It makes it hard for me to track the increment changes you are making especially in big PRs like this...

File: src/db-backup.ts [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+  return {
+    name: backUpName,
+    block: { number: new BigNumber(blockNumber), hash }
+  }
+}
+
+export function getBackUps (): BackUpEntry[] {
+  const backupConfig = config.get<DbBackUpConfig>('dbBackUp')
+
+  const backups = fs.readdirSync(path.resolve(process.cwd(), `./${backupConfig.path}`))
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/db-backup.ts:26]: Done. Thanks for pointing

File: src/db-backup.ts [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+  readonly backUpConfig: DbBackUpConfig
+
+  constructor (eth: Eth) {
+    if (!config.has('dbBackUp')) {
+      throw new Error('DB Backup config not exist')
+    }
+    this.backUpConfig = config.get<DbBackUpConfig>('dbBackUp')
+    this.db = config.get<string>('db')
+
+    if (!fs.existsSync(this.backUpConfig.path)) {
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/db-backup.ts:53]: Done

File: test/cli.spec.ts (hunk 10 above)
Comment (nduchak) [reply test/cli.spec.ts:41]: I think we can remove it

File: src/cli/start.ts (hunk 12 above)
Comment (nduchak) [reply src/cli/start.ts:127]: Yes, it's true. For that case, we should notify DevOps that the restore process fails

File: src/db-backup.ts [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+  public async restoreDb (): Promise<void> {
+    const backUps = getBackUps()
+    const [_, oldest] = backUps
+
+    if (backUps.length < 2) {
+      throw new Error('Should be two backups to be able to restore')
+    }
+
+    // Check if back up block hash exist after reorg
+    const block = await this.eth.getBlock(oldest.block.hash).catch(() => false)
```
Comment (AuHau) [inline src/db-backup.ts:93]: I believe this is not what we want. You should fetch the block based on its block number that you have stored in the backup's name and then compare the block's hash with the hash you have stored in the backup's filename. If they match, great we have a valid backup if not, then you know that the reorg is bigger than the backup and then we are screwed 😅

Comment (AuHau) [reply src/db-backup.ts:93]: Ok, this is actually valid approach as well.

File: src/cli/start.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nduchak) [reply src/cli/start.ts:113]: Reverted back

**Review by AuHau — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**